### DISCUSSION TITLE
vanilla js implementation | No Framework

### DISCUSSION THREAD

**Discussion by EugenePeter** [TanStack/table#4928] [General] [answered]
Is there an existing example of implementing table using the @tanstack/table-core.
At the moment docs only have:

`import { createTable } from '@tanstack/table-core'

const table = createTable(options)`

I tried to implement like so:

`  const columnHelper = createColumnHelper<Person>();
  const columns = [
    columnHelper.accessor("firstName", {
      cell: (info) => info.getValue(),
      footer: (info) => info.column.id,
    }),
    columnHelper.accessor((row) => row.lastName, {
      id: "lastName",
      cell: (info) => <i>{info.getValue()}</i>,
      header: () => <span>Last Name</span>,
      footer: (info) => info.column.id,
    }),
    columnHelper.accessor("age", {
      header: () => "Age",
      cell: (info) => info.renderValue(),
      footer: (info) => info.column.id,
    }),
    columnHelper.accessor("visits", {
      header: () => <span>Visits</span>,
      footer: (info) => info.column.id,
    }),
    columnHelper.accessor("status", {
      header: "Status",
      footer: (info) => info.column.id,
    }),
    columnHelper.accessor("progress", {
      header: "Profile Progress",
      footer: (info) => info.column.id,
    }),
  ];

  const data: Person[] = [
    {
      firstName: "tanner",
      lastName: "linsley",
      age: 24,
      visits: 100,
      status: "In Relationship",
      progress: 50,
    },
    {
      firstName: "tandy",
      lastName: "miller",
      age: 40,
      visits: 40,
      status: "Single",
      progress: 80,
    },
    {
      firstName: "joe",
      lastName: "dirte",
      age: 45,
      visits: 20,
      status: "Complicated",
      progress: 10,
    },
  ];

  const table = createTable({
    columns,
    data,
    getCoreRowModel: getCoreRowModel(),
    state: {}, // Provide the state property here
    onStateChange: () => {}, // Provide the onStateChange property here
    renderFallbackValue: () => null, // Provide the renderFallbackValue property here
  });`
  
  `      <table>
        <thead>
          {
            table?.getHeaderGroups()?.map((headerGroup) => (
              <tr key={headerGroup.id}>
                {headerGroup?.headers?.map((header) => (
                  <th key={header.id}>
                    hello
                    {/* {header.isPlaceholder
                        ? null
                        : flexRender(
                            header.column.columnDef.header,
                            header.getContext()
                          )} */}
                  </th>
                ))
                }
              </tr>
            ))
          }
          hello
        </thead>
        <tbody>
          {table?.getRowModel()?.rows?.map((row) => (
            <tr key={row.id}>
              {row.getVisibleCells().map((cell) => {
                const value = cell?.value;
                console.log("VALUEEEEE:", value);
                return <td key={cell.id}>hello</td>;
                return (
                  <td key={cell.id}>
                    {/* {flexRender(cell.column.columnDef.cell, cell.getContext())} */}
                  </td>
                );
              })}
            </tr>
          ))}
        </tbody>
      </table>`
      
      Notice that I commented out flexRenderer, I don't if it's available from '@tanstack/table-core'
      
      Now I am getting error:
      `Cannot read properties of undefined (reading 'left')`
      
      `    at file:///C:/Users/epmae/Desktop/eugene/development/qwik-test/node_modules/@tanstack/table-core/build/lib/index.mjs:1738:97
    at Object.getLeftVisibleCells (file:///C:/Users/epmae/Desktop/eugene/development/qwik-test/node_modules/@tanstack/table-core/build/lib/index.mjs:59:21)
    at getVisibleCells.memo.key (file:///C:/Users/epmae/Desktop/eugene/development/qwik-test/node_modules/@tanstack/table-core/build/lib/index.mjs:2559:40)
    at Object.getVisibleCells (file:///C:/Users/epmae/Desktop/eugene/development/qwik-test/node_modules/@tanstack/table-core/build/lib/index.mjs:59:21)
    at C:\Users\epmae\Desktop\eugene\development\qwik-test\src\routes\members\index.tsx:115:20
    at Array.map (<anonymous>)
    at AsyncFunction.members_component_DpttA29SYVY (C:\Users\epmae\Desktop\eugene\development\qwik-test\src\routes\members\index.tsx:113:40)
    at AsyncFunction.invoke (C:\Users\epmae\Desktop\eugene\development\qwik-test\dist-dev\tsc-out\packages\qwik\src\core\util\implicit_dollar.js:1290:26)
    at C:\Users\epmae\Desktop\eugene\development\qwik-test\dist-dev\tsc-out\packages\qwik\src\core\util\implicit_dollar.js:8158:35
    at then (C:\Users\epmae\Desktop\eugene\development\qwik-test\dist-dev\tsc-out\packages\qwik\src\core\util\implicit_dollar.js:433:56`

**Comment by KasimAhmic** [comment]
Did you ever figure this out? Running into the same issue now.

**Reply by marceloverdijk** [reply]
I tried once for a POC but abandoned as there is no documentation or further assistance unfortunately.

Once somebody figured out I might give Tanstack Table a retry in the future.

**Reply by marceloverdijk** [reply]
I created this repo in the past to come to a working solution:

https://github.com/marceloverdijk/astro-tanstack-table-vanilla

Had a brief discussion with @tannerlinsley on Twitter about it but unfortunately never emerged into something useful. There seems not to be enough interest, nor examples online. I even wondered back then if a no-framework approach is feasible, despite being advertised.

**Reply by KasimAhmic** [reply]
Ah bummer. Oh well, thanks anyway for the reply!

**Comment by geries-visdom** [comment]
For those landing here you need to have a proper `state`  with minimal `props` like this 

```
 const state = {
    columnPinning: {},
    pagination: {
      pageIndex: 0,
      pageSize: 20,
    },
  };
  const table = createTable<Person>({
    data: defaultData,
    columns,
    getCoreRowModel: getCoreRowModel(),
    getPaginationRowModel: getPaginationRowModel(),
    state,
    onStateChange: () => {}, // noop
    renderFallbackValue: undefined,
    debugAll: true,
  });
```

The should solve the exception you getting in the `table` object.

Then when doing the actual rendering, you need to do this:

First you need a `flexRenderer` which will make it easier to fingure out if the value is a function(called template by them) or a string

````
const flexRenderer = (comp, attrs) => {
  if (typeof comp === "function") {
    return comp(attrs);
  }

  return comp
};

````

Then it is used like this when rendering the table

```
 header.isPlaceholder ? "" : flexRenderer(header.column.columnDef?.header, header.getContext()),
// or in case of the cell
flexRenderer(cell.column.columnDef?.cell, cell.getContext())
```

**Reply by marceloverdijk** [reply]
Thx Geries!, I tried it last weekend and I was able to construct the table and draw it.

It's worth to mention that even when not using the column pinning feature of Tanstack Table the `columnPinning: {},` in the `state` definition is needed. I think these kind of nuances caused many of stumbling on this to be blocked.

Next weekend I will continue with my experiment and try to make it interactive: clicking next page button to re-render the table using a custom js implementation. And after that sorting is on my list.
I'm wondering if you already did something like this and if you are willing to share you experience?

**Reply by geries-visdom** [reply]
@marceloverdijk nop sorry we moved on to other projects and felt like overhead to try to use this with `mithril.js`

**Reply by marceloverdijk** [reply]
No worries , I'm making progress myself and will post a full example here soon.

**Reply by marceloverdijk** [reply]
I made this example on StackBlitz:

https://stackblitz.com/edit/github-6xqt97?file=main.js

It nicely renders the table, but pagination is not working unfortunately yet, but it can be a good start for others interested.
The `table.nextPAge()` and `table.previousPage()` calls do not seem to update the `table` state at all...
I also create a separate issue for this: https://github.com/TanStack/table/issues/5375

**Reply by marceloverdijk** [reply]
This thread on X with @KevinVandy is also interesting to follow: https://twitter.com/marceloverdijk/status/1762606611535687699

Apparently when using Vanilla JS we should not use `state` but `initialState` and remove the `onStateChange`...
However this gives new problems: https://stackblitz.com/edit/github-6xqt97-4ufpds?file=main.js

```
Uncaught TypeError: Cannot read properties of undefined (reading 'columnOrder')
    at table._getOrderColumnsFn.orderedColumns (Ordering.ts:114:25)
```

I hope @KevinVandy might chime in here to help and give some pointers to keep going, as I'm stuck now and out of ideas how to continue.

**Reply by KevinVandy** [reply]
No, not true. I hadn't realized that they were using vanilla js. Thought that they were using React.

**Reply by marceloverdijk** [reply]
Thx @KevinVandy for stepping in.

In this topic we are trying to indeed get Tanstack table running with vanilla js without any framework (so indeed not using React).

So from your last comment do I understand correct the usage of `state` and that noop `onStateChange: () => {}` is correct?

Do you have any other pointers on how to get it working. In https://stackblitz.com/edit/github-6xqt97?file=main.js `nextPage()`/`previousPage()` will not change the state..

I'm a little bit confused now, what behaviour is expected, and e.g. if the comment in https://github.com/TanStack/table/issues/5375 is correct then.

**Reply by marceloverdijk** [reply]
I came a bit further this evening, and I have a table with working pagination now:

https://github.com/marceloverdijk/astro-tanstack-table-vanilla
https://stackblitz.com/github/marceloverdijk/astro-tanstack-table-vanilla

I'm using [Nanostores Atom](https://github.com/nanostores/nanostores) for state management.
I basically followed the React adapter and changed it to use Nanostores.
Not sure if this is the best approach or if it has drawbacks... love to get some feedback on it.

**Reply by fsmeier** [reply]
Exceptions because of not setting default values in vanilla like`const state = {
    columnPinning: {},
    pagination: {
      pageIndex: 0,
      pageSize: 20,
    },
  };` seems like a bug to me...

**Comment by rlsystems** [comment]
I couldn't even figure out how to import the library, would love a full example. Trying to get this working in a simple html template for now. Ultimately in an ASP.NET Razor Pages project

**Reply by KasimAhmic** [reply]
I've since given up on this but I did go back and try to get it working for fun. Wasn't able to get very far, but I did manage to put this together. No clue how to actually get the table to render on the page but this is at least constructing the table object so it's a start.

To import the library, I used [unpkg](https://unpkg.com) to get the UMD build and then I run the `renderTable` function once it loads. You can access `@tanstack/table-core` using the global `TableCore` object once the library is loaded.

Hope this helps!

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vanilla JS TanStack Table</title>

    <script
      src="https://unpkg.com/@tanstack/table-core@8.11.3/build/umd/index.production.js"
      onload="renderTable()"
      defer
    ></script>

    <script>
      function renderTable() {
        const columnHelper = TableCore.createColumnHelper();

        const state = {
          columnPinning: {},
          pagination: {
            pageIndex: 0,
            pageSize: 20,
          },
        };

        const data = [
          {
            firstName: 'tanner',
            lastName: 'linsley',
            age: 24,
            visits: 100,
            status: 'In Relationship',
            progress: 50,
          },
          {
            firstName: 'tandy',
            lastName: 'miller',
            age: 40,
            visits: 40,
            status: 'Single',
            progress: 80,
          },
          {
            firstName: 'joe',
            lastName: 'dirte',
            age: 45,
            visits: 20,
            status: 'Complicated',
            progress: 10,
          },
        ];

        const columns = [
          columnHelper.accessor('firstName', {
            cell: (info) => info.getValue(),
            footer: (info) => info.column.id,
          }),
          columnHelper.accessor((row) => row.lastName, {
            id: 'lastName',
            cell: (info) => {
              info.getValue();
            },
            header: () => 'Last Name',
            footer: (info) => info.column.id,
          }),
          columnHelper.accessor('age', {
            header: () => 'Age',
            cell: (info) => info.renderValue(),
            footer: (info) => info.column.id,
          }),
          columnHelper.accessor('visits', {
            header: () => Visits,
            footer: (info) => info.column.id,
          }),
          columnHelper.accessor('status', {
            header: 'Status',
            footer: (info) => info.column.id,
          }),
          columnHelper.accessor('progress', {
            header: 'Profile Progress',
            footer: (info) => info.column.id,
          }),
        ];

        const table = TableCore.createTable({
          data,
          columns,
          getCoreRowModel: TableCore.getCoreRowModel(),
          getPaginationRowModel: TableCore.getPaginationRowModel(),
          state,
          onStateChange: () => {}, // noop
          renderFallbackValue: undefined,
          debugAll: true,
        });

        console.log(table);
      }
    </script>
  </head>
  <body>
    <h1>Vanilla JS TanStack Table</h1>
  </body>
</html>
```

**Reply by marceloverdijk** [reply]
@KasimAhmic I tried this last weekend as well. Did you already experiment further to make the table interactive, like pagination?

**Reply by KasimAhmic** [reply]
@marceloverdijk Had some time to kill this evening and I gave pagination another shot. Using your example from your conversation with @KevinVandy as a starting point, I was able to get pagination working though I'm unsure if if this is the "right" way to do it.

I use `table.nextPage()` and `table.previousPage()` to trigger a state update for which I listen to using the `onStateChange` table option. Then I compute the new state by passing the current state retrieved via `table.getState()` to the `onStateChange`'s updater function argument, set `table.options.state` to that computed value, and finally call `renderTable()` which blows away the contents of the table container and re-renders it.

```typescript
    onStateChange: (updaterFn) => {
      table.options.state = updaterFn(table.getState()); // Unsure if this is okay to do, but it seems to work

      renderTable();
    },
```

You can see the full example here: https://stackblitz.com/edit/github-6xqt97-uitzku?file=main.js

**Reply by marceloverdijk** [reply]
Thx @KasimAhmic this seems much simpler indeed, and no need for Nanostores or any other state management library or implementation. 
It almost sounds to simple ;-) would love to see some feedback from @KevinVandy or @tannerlinsley on this whole discussion, but they unfortunately keep quiet.

**Reply by KevinVandy** [reply]
Sorry, I use frameworks, not vanilla js. Happy to see people discovering usage with vanilla. Consider contributing your findings to the docs. I can help organize that.

**Reply by marceloverdijk** [reply]
ok thx @KevinVandy for the clarification.
@KasimAhmic approach is very simple and seems to work (at least with basic setup including pagination).
If @tannerlinsley could confirm this approach is good, then I'm willing to contribute some Vanilla examples to the docs.
But first I would like confirmation that this is a good approach, before spending time on these examples. I hope you understand.

**Reply by marceloverdijk** [reply]
This evening I went further with @KasimAhmic example and have created 2 examples on StackBlitz:

- [Tanstack Table Vanilla Example: Basic](https://stackblitz.com/edit/vitejs-vite-mcxo6f?file=src%2Fmain.ts)
- [Tanstack Table Vanilla Example: Pagination](https://stackblitz.com/edit/vitejs-vite-zbmszd?file=src%2Fmain.ts)

The basically mimic the React Basic and pagination examples.
So the first one is maybe not so interesting as there is not much happening, but the second one with pagination is completely interactive using Vanilla TS.

The create and render parts are completely independent of each other and thus decoupled.

@KevinVandy if you can help organising to contribute this to the project that would be great.

Maybe the `createVanillaTable` method (maybe not the best name) can be added as adapter to the project, together with providing some examples similar as is done for the other adapters. What do you think?

**Reply by marceloverdijk** [reply]
I have even improved the `createVanillaTable` adapter implementation in my local project.
It allows partial state to be provided in the create call, similar as the other adapters do, which makes it more easier to use for the user.

This also means e.g. the `columnPinning: {},` is not needed to supply anymore, which seemed quite strange to be needed.

https://stackblitz.com/edit/vitejs-vite-wcxv2m?file=src%2Fcreate-vanilla-table.ts

**Reply by geries-visdom** [reply]
Wow this is awesome work.

**Reply by 0Chan-smc** [reply]
@marceloverdijk Your work really helped me a lot. Thank you so much.

**Reply by KevinVandy** [reply]
would really like to get these vanilla examples in the official docs @marceloverdijk. A PR from you would speed this up, otherwise, I'll eventually copy these suggestions.

**Reply by mario-yanill** [reply]
In this approach how do I update the table content (rows)? If anyone could help it would be great, thanks in advance

**Reply by 0Chan-smc** [reply]
@mario-yanill  Have you tried creating another table with createVanillaTable with new content in it? You can then simply render it.

**Comment by fsmeier** [comment]
+1 for vanilla examples

**Reply by marceloverdijk** [reply]
Sorry, for not earlier responding.
But actually, today I was looking into a vanilla implementation again during the holidays ;-)

However, like in the https://stackblitz.com/github/marceloverdijk/astro-tanstack-table-vanilla example, creating the html manually is quite painful... so today I worked on a proof of concept with `@tanstack/table-core`, `nanostores` for state management, and `lit-html` for html templates.
Note not to confuse `lit-html` with `lit-element` and the Tanstack Table `lit` adapter. 

The outcome was I had searchable, sortable and pageable page with vanilla js/ts.

Here is a stackblitz example of it:

https://stackblitz.com/edit/github-isvmcumr?file=src%2Fpages%2Findex.astro


Back to the vanilla examples, @KevinVandy do you think it's a good idea to create vanilla examples with `lit-html` for html templating?
Let me know, I will make it one of my new years resolutions to provide examples!

On a related note `@tanstack/table-core` provides the bare bones `createTable` function, and for frameworks there are `use*` variants with `flexRender` functions. I think for vanilla js it could be useful to have a `use*` and `flexRender` as well.

for example as simple as:

```
export function flexRender<TProps extends object>(comp: any, props: TProps) {
  if (typeof comp === 'function') {
    return comp(props);
  }
  return comp;
}
```

What do you think?

**Reply by 0Chan-smc** [reply]
I prefer to use fewer dependencies. But the example with `lit-html` is not bad.

**Reply by KevinVandy** [reply]
We already have a Lit adapter, so if someone is going to use Lit, they should use these. https://tanstack.com/table/latest/docs/framework/lit/examples/filters

I think a vanilla example should define it's own flexRender function?

**Reply by marceloverdijk** [reply]
lit-html is a standalone html template renderer, it's not the same as lit element with controllers etc as in the lit adapter examples.

But no problem, the vanilla examples can be done without lit-html. At the end everybody can choose what ever they want for the html rendering.

**Comment by marceloverdijk** [comment]
Here we go, a first take at the PR for a basic and pagination example using Vanilla js/ts (without any dependency except nanostores):
https://github.com/TanStack/table/pull/5853

@KevinVandy let me know what you think

**Comment by marceloverdijk** [comment, marked as answer]
Vanilla examples basic, pagination and sorting are now available here:

- https://tanstack.com/table/latest/docs/framework/vanilla/examples/basic
- https://tanstack.com/table/latest/docs/framework/vanilla/examples/pagination
- https://tanstack.com/table/latest/docs/framework/vanilla/examples/sorting

Let's mark this discussion thread as answered.
